connectors/memory/memory.go

```diff
@@ -0,0 +1,493 @@
+// Otherwise, search the partition for the exact same clustering keys. If there, fail
+// if not, then insert it at the right spot (sort.Search does most of the heavy lifting here)
+func (c *Connector) CreateIfNotExists(_ context.Context, ei *dosa.EntityInfo, values map[string]dosa.FieldValue) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+	if c.data == nil {
+		c.data = make(map[string]map[string][]map[string]dosa.FieldValue)
+	}
+	if c.data[ei.Def.Name] == nil {
+		c.data[ei.Def.Name] = make(map[string][]map[string]dosa.FieldValue)
+	}
+	entityRef := c.data[ei.Def.Name]
+	encodedPartitionKey := partitionKeyBuilder(ei, values)
+	if entityRef[encodedPartitionKey] == nil {
+		entityRef[encodedPartitionKey] = make([]map[string]dosa.FieldValue, 0, 1)
+	}
+	partitionRef := entityRef[encodedPartitionKey]
+	// no data in this partition? easy out!
+	if len(partitionRef) == 0 {
+		entityRef[encodedPartitionKey] = append(entityRef[encodedPartitionKey], values)
+		return nil
+	}
+
+	if len(ei.Def.ClusteringKeySet()) == 0 {
+		// no clustering key, so the row must already exist
+		return &dosa.ErrAlreadyExists{}
+	}
+	// there is a clustering key, find the insertion point (binary search would be fastest)
+	found, offset := findInsertionPoint(ei, partitionRef, values)
+	if found {
+		return &dosa.ErrAlreadyExists{}
+	}
+	// perform slice magic to insert value at given offset
+	l := len(entityRef[encodedPartitionKey])                                                                     // get length
+	entityRef[encodedPartitionKey] = append(entityRef[encodedPartitionKey], entityRef[encodedPartitionKey][l-1]) // copy last element
+	// scoot over remaining elements
+	copy(entityRef[encodedPartitionKey][offset+1:], entityRef[encodedPartitionKey][offset:])
+	// and plunk value into appropriate location
+	entityRef[encodedPartitionKey][offset] = values
+
+	return nil
+}
+
+// Read searches for a row. First, it finds the partition, then it searches in the partition for
+// the data, and returns it when it finds it. Again, sort.Search does most of the heavy lifting
+// within a partition
+func (c *Connector) Read(_ context.Context, ei *dosa.EntityInfo, values map[string]dosa.FieldValue, fieldsToRead []string) (map[string]dosa.FieldValue, error) {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	if c.data[ei.Def.Name] == nil {
+		return nil, &dosa.ErrNotFound{}
+	}
+	entityRef := c.data[ei.Def.Name]
+	encodedPartitionKey := partitionKeyBuilder(ei, values)
+	partitionRef := entityRef[encodedPartitionKey]
+	// no data in this partition? easy out!
+	if len(partitionRef) == 0 {
+		return nil, &dosa.ErrNotFound{}
+	}
+
+	if len(ei.Def.ClusteringKeySet()) == 0 {
+		return filterSet(partitionRef[0], fieldsToRead), nil
+	}
+	// clustering key, search for the value in the set
+	found, inx := findInsertionPoint(ei, partitionRef, values)
+	if !found {
+		return nil, &dosa.ErrNotFound{}
+	}
+	return filterSet(partitionRef[inx], fieldsToRead), nil
```

Comment:
I don't think that's possible. filterSet only filters fields. I guess it's possible to return no fields, but there's still a row there, and so it should do the right thing